DOCSP-11463 (#401)
mongodb/docs-realm: source/android/open-a-realm.txt

.. _android-open-realm:

======================
Open and Close a Realm
======================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> Open a Synced Realm

Reviewer: We should definitely add a note about enabling sync via gradle to use a synced realm (see the "enable sync" section of this step in the install guide: https://docs.mongodb.com/realm/android/install/#add-realm-to-the-application-level-build-gradle-file). Might be a good idea to add a ref to that section of the step and link to it so we don't have to duplicate that text/update it when the SDK changes its mind in the future.

-------------------

> To open a synced realm, you call 

Reviewer 2: ```suggestion
To open a synced realm, call 
```

> To open a synced realm, you call 

Reviewer: ```suggestion
To open a synced realm, call 
```
(closer to the convention we've used for other examples)

:java-sdk:`getInstanceAsync()
</io/realm/Realm.html#getInstanceAsync-io.realm.RealmConfiguration-io.realm.Realm.Callback->`, 
passing in a a :java-sdk:`SyncConfiguration <io/realm/mongodb/sync/SyncConfiguration.html>` 

> object. For example, the following code demonstrates how to create a realm with 

Reviewer: You should probably replace all occurrences of "realm" in the text of your PR with `{+realm+}` since that's the norm elsewhere in the repo.

Author: good catch!

specific sync settings created using a ``SyncConfiguration`` object:

.. tabs-realm-languages::
   
   .. tab::
      :tabid: kotlin
      
      .. code-block:: kotlin

>          val user: User? = app.currentUser()

Reviewer: I would probably add the `App` instantiation to the Kotlin example as well for parity with the java example:

```
val appID = "<your app ID>" // replace this with your App ID
```

OR you could ignore this and follow my suggestion in my comment on the java block.

         val partitionValue: String = "myPartition"
         val config = 
         SyncConfiguration.Builder(user!!, partitionValue).build()

         val realm: Realm = Realm.getInstance(config)

   .. tab::
      :tabid: java
      
      .. code-block:: java

>          App app = new App("app-id");

Reviewer: ```suggestion
         String appID = "<your app ID>"; // replace this with your App ID
```

... or alternatively you could remove the App line from the java section and not add it to the kotlin section -- that way we can assume that the App is already instantiated and the User has already logged in. As it stands, the user never has a chance to log in in the java example which I imagine will cause some reader confusion.

Author: went with this choice. :)

         User user = app.currentUser();
         String partitionValue = "myPartition";
         SyncConfiguration config = 
            new SyncConfiguration.Builder(user, partitionValue).build();

         Realm realm = Realm.getInstance(config);
   

> In the example above, the code is showing how to open the realm *asynchronously* 

Reviewer: ```suggestion
In the example above, the code shows how to open the realm *asynchronously* 
```

by using :java-sdk:`getInstanceAsync()
</io/realm/Realm.html#getInstanceAsync-io.realm.RealmConfiguration-io.realm.Realm.Callback->`. 

You can also open a realm synchronously by using :java-sdk:`getInstance()
<io/realm/Realm.html#getInstance-io.realm.RealmConfiguration->`, but this may 
lead to temporary data inconsistencies while the remote data is downloaded, and 

> is generally not recommended. 

Reviewer: I think the section here after "but" might be better suited to an "admonition" -- what do you think?

Author: Don't want to highlight it or call attention to it, since it's not a best practice.

Open a Local (Non-Synced) Realm
-------------------------------

> When opening a local (non-synced) realm, you set the default settings and then 

Reviewer: ```suggestion
When opening a local (non-synced) realm, set the default settings and then 
```

.. tabs-realm-languages::
   
   .. tab::
      :tabid: kotlin
      
      .. code-block:: kotlin

         val user: User? = app.currentUser()
         val partitionValue: String = "myPartition"
         val config = SyncConfiguration.Builder(user!!, partitionValue).build()

         val realm: Realm = Realm.getInstance(config)

   .. tab::
      :tabid: java
      
      .. code-block:: java

         User user = app.currentUser();
         String partitionValue = "myPartition";
         SyncConfiguration config = 
            new SyncConfiguration.Builder(user, partitionValue).build();

         Realm realm = Realm.getInstance(config);
   

The code above shows how to open the {+realm+} *asynchronously* 
by using :java-sdk:`getInstanceAsync()
</io/realm/Realm.html#getInstanceAsync-io.realm.RealmConfiguration-io.realm.Realm.Callback->`. 
You can also open a {+realm+} synchronously by using :java-sdk:`getInstance()
<io/realm/Realm.html#getInstance-io.realm.RealmConfiguration->`, but this may 
lead to temporary data inconsistencies while the remote data is downloaded, and 
is generally not recommended. 

.. note::
   
   For more information on enabling sync in Android, see 

>    :ref:`Install Realm for Android <android-install>`.

Reviewer: I think pointing users to this section is more helpful than what we had before, but I fear that the section might be so long and verbose that many users will tune out the sync:enabled section (thinking that this page just tells them how to pull down the dependency). If we can't create a ref to the sync:enabled section of the install guide, it might be a good idea to just duplicate the content in this note and link to the install android section "for more information."

Open a Local (Non-Synced) Realm
-------------------------------
When opening a local (non-synced) {+realm+}, set the default settings and then 
call :java-sdk:`getDefaultInstance()<io/realm/Realm.html#getDefaultInstance-->`:

.. tabs-realm-languages::
   
   .. tab::
      :tabid: kotlin
      
      .. code-block:: kotlin

         val config = RealmConfiguration.Builder().build()
         Realm.setDefaultConfiguration(config)

         val realm: Realm = Realm.getDefaultInstance()

   .. tab::
      :tabid: java
      
      .. code-block:: java

         RealmConfiguration realmConfig = 
            new RealmConfiguration.Builder(this).build();
         Realm.setDefaultConfiguration(realmConfig);

         Realm realm = Realm.getDefaultInstance();

Close a Realm
-------------
It is important to remember to call the ``close()`` method when done with a 

> {+realm+} instance. Failing to do so can lead to OutOfMemoryError as the native 

Reviewer: Should probably monospace OutOfMemoryError
